### PR TITLE
Clean up ex-malloc() use of CHIPMem.h

### PR DESCRIPTION
**Pull request by kpschoedel**
#### Problem

As part of heap memory management, a recent change (#3143) replaced
malloc()-family calls one-for-one with their "CHIPMem.h" equivalents,
but a one-to-one replacement is not always ideal.

#### Summary of Changes

- Use ScopedMemoryBuffer instead where possible.
- Added ScopedMemoryString, since this is a common case.
- Some additional safety checks.

### CODE REVIEW

File: src/inet/tests/TestInetCommon.cpp [hunk 1]
```diff
@@ -290,8 +290,8 @@ void InitNetwork()
         {
             uint64_t iid    = gNetworkOptions.LocalIPv6Addr[j].InterfaceId();
             char * tap_name = (char *) chip::Platform::MemoryAlloc(sizeof(gDefaultTapDeviceName));
+            assert(tap_name);
```
Comment (gerickson) [inline src/inet/tests/TestInetCommon.cpp:293]: This seems unrelated to this check-in.

File: src/lib/support/CHIPMemString.h [hunk 2]
```diff
@@ -34,18 +36,69 @@ namespace Platform {
...
+     *                              Like `strndup()` but unlike `strncpy()`, the result is always
+     *                              null-terminated.
+     */
+    ScopedMemoryString(const char * string, size_t length = 0)
+    {
+        if (string && !length)
+        {
+            length = strlen(string);
+        }
+        if (string || length)
```
Comment (gerickson) [inline src/lib/support/CHIPMemString.h:96]: If string is NULL but length is non-zero, is there still a string to copy?

File: src/inet/tests/TestInetCommon.cpp (hunk 1 above)
Comment (kpschoedel) [reply src/inet/tests/TestInetCommon.cpp]: Split to https://github.com/project-chip/connectedhomeip/pull/3230

File: src/lib/support/CHIPMemString.cpp [hunk 3]
```diff
@@ -31,17 +31,21 @@ namespace Platform {
 
 char * MemoryAllocString(const char * string, size_t length)
 {
-    if (length == 0)
+    if (string && !length)
     {
         length = strlen(string);
```
Comment (tcarmelveilleux) [inline src/lib/support/CHIPMemString.cpp:36]: This can overrun the end. Are there many uses of `length=0` in calls? Strongly suggest avoiding this particular usage of len==0, unless there is a sane max length provided.

File: src/lib/support/CHIPMemString.h (hunk 2 above)
Comment (kpschoedel) [reply src/lib/support/CHIPMemString.h:96]: The effect is as if the string is empty. I can remove this if preferred.

File: src/lib/support/CHIPMemString.cpp (hunk 3 above)
Comment (rwalker-apple) [reply src/lib/support/CHIPMemString.cpp:36]: I find the contract odd.  Is this supposed to be something like a `strndup()`?

Comment (kpschoedel) [reply src/lib/support/CHIPMemString.cpp:36]: Only one caller actually supplies a length — all the rest were previously `strdup()` —and would just change from `MemoryAlloc(s)` to `MemoryAlloc(s, strlen(s))`. The allocated size is always `length+1`.

Comment (kpschoedel) [reply src/lib/support/CHIPMemString.cpp:36]: > I find the contract odd. Is this supposed to be something like a `strndup()`?

Yes, this replaced str[n]dup() calls.

Comment (rwalker-apple) [reply src/lib/support/CHIPMemString.cpp:36]: this does not replicate `strndup()`...

```
#include <stdio.h>
#include <string.h>

int main()
{
    char * p = strndup("hi", 0);
    printf("p = %p\n", p);
    if (p)
    {
        printf("strlen(p) = %lu\n", strlen(p));
    }
}
```
yields:
```
gcc -o strndup strndup.c && ./strndup
p = 0x7fd8e6c058f0
strlen(p) = 0


```

Comment (rwalker-apple) [reply src/lib/support/CHIPMemString.cpp:36]: other aspects of the strndup() contract not met with this change:
* `strndup(null, 0)` returns the empty string
* `strndup(null, 1)` aborts

Comment (kpschoedel) [reply src/lib/support/CHIPMemString.cpp:36]: I'll replace the default parameter with an overload to match the 0 case. Passing NULL to str* is UB, though, and I'd rather provide defined behaviour — the current choice is to treat a null source as an empty string; would you prefer an abort()?

Comment (rwalker-apple) [reply src/lib/support/CHIPMemString.cpp:36]: blind dereference of the pointer would be my preference.  extra behavior is more to document, harder to debug.

Comment (tcarmelveilleux) [reply src/lib/support/CHIPMemString.cpp]: My main concern is the fact that this may cause possibly very large allocations on passing-in of garbage or carefully crafted data. I feel like `strdup()` is always incorrect in embedded, and in fact, I have worked on many coding guidelines that forbade the "n"-less string functions (e.g. strlen, strdup, strcpy) and mandated strnlen, strndup, strncpy/strlcpy, etc. I feel like this "convenience" string duplicate with no upper bounds provided by caller may not be desirable on constrained devices.

Comment (rwalker-apple) [reply src/lib/support/CHIPMemString.cpp]: much larger change to keep folks from calling strlen() on the interrupt vector... ;)

Comment (kpschoedel) [reply src/lib/support/CHIPMemString.cpp]: OK. I guess it'll make it a little easier on someone later just to hunt down all the `strlen`s.

File: src/lib/core/tests/TestCHIPTLV.cpp [hunk 4]
```diff
@@ -230,14 +230,13 @@ void TestDupString(nlTestSuite * inSuite, TLVReader & reader, uint64_t tag, cons
     uint32_t expectedLen = strlen(expectedVal);
     NL_TEST_ASSERT(inSuite, reader.GetLength() == expectedLen);
 
-    char * val = static_cast<char *>(chip::Platform::MemoryAlloc(expectedLen + 1));
+    chip::Platform::ScopedMemoryBuffer<char> valBuffer;
+    char * val = static_cast<char *>(valBuffer.Alloc(expectedLen + 1).Get());
```
Comment (andy31415) [inline src/lib/core/tests/TestCHIPTLV.cpp:234]: is the cast needed? the scoped buffer template should already be typed.

same for below

File: src/lib/support/CHIPMemString.h [hunk 5]
```diff
@@ -21,31 +21,79 @@
...
+ *
+ * @param[in]  length           Length to be copied.
+ *
+ * @retval  Same as `dest`.
+ */
+inline char * CopyString(char * dest, const char * source, size_t length)
+{
+    if (dest)
+    {
+        memcpy(dest, source, length);
```
Comment (andy31415) [inline src/lib/support/CHIPMemString.h:54]: any issues if source is shorter than length? Maybe we could use strncpy internally.

**Review by andy31415 — APPROVED** [review]

File: src/lib/core/tests/TestCHIPTLV.cpp (hunk 4 above)
Comment (kpschoedel) [reply src/lib/core/tests/TestCHIPTLV.cpp]: Right, was left over from Platform::MemoryAlloc.

File: src/lib/support/CHIPMemString.h (hunk 5 above)
Comment (kpschoedel) [reply src/lib/support/CHIPMemString.h]: I was about to say no (NUL in the source is fine, NUL not in the source is fine) but there is one case where memcpy would break (NUL-terminated source closely followed by dangerous addresses).

**Review by rwalker-apple — APPROVED** [review]

**Review by BroderickCarlin — APPROVED** [review]
